service/handler.go

```diff
@@ -1,8 +1,175 @@
+		http.Error(w, "failed to decode request body", http.StatusBadRequest)
+		return
+	}
+	defer r.Body.Close()
+
+	// Decrypt data
+	encryptedDataBz, err := base64.StdEncoding.DecodeString(reqBody.EncryptedDataBase64)
+	if err != nil {
+		log.Errorf("failed to decode encrypted data: %s", err.Error())
+		http.Error(w, "failed to decode encrypted data", http.StatusBadRequest)
+		return
+	}
+
+	providerAcc, err := queryClient.GetAccount(reqBody.ProviderAddress)
+	if err != nil {
+		log.Errorf("failed to get provider's account: %s", err.Error())
+		http.Error(w, "failed to get provider's account", http.StatusBadRequest)
+		return
+	}
+
+	if providerAcc.GetPubKey() == nil {
+		log.Errorf("failed to get public key of provider's account: %s", err.Error())
+		http.Error(w, "failed to get public key of provider's account", http.StatusBadRequest)
+		return
+	}
+
+	providerPubKeyBytes := providerAcc.GetPubKey().Bytes()
+
+	providerPubKey, err := btcec.ParsePubKey(providerPubKeyBytes, btcec.S256())
+	if err != nil {
+		log.Errorf("failed to parse provider's public key: %s", err.Error())
+		http.Error(w, "failed to parse provider's public key", http.StatusBadRequest)
+		return
+	}
+
+	decryptSharedKey := crypto.DeriveSharedKey(oraclePrivKey, providerPubKey, crypto.KDFSHA256)
+
+	decryptedData, err := crypto.Decrypt(decryptSharedKey, nil, encryptedDataBz)
+	if err != nil {
+		log.Errorf("failed to decrypt data: %s", err.Error())
+		http.Error(w, "failed to decrypt data", http.StatusBadRequest)
+		return
+	}
+
+	// Validate data
+	dataHash := sha256.Sum256(decryptedData)
+	dataHashStr := hex.EncodeToString(dataHash[:])
+	if reqBody.DataHash != dataHashStr {
+		log.Errorf("data hash mismatch")
+		http.Error(w, "data hash mismatch", http.StatusBadRequest)
+		return
+	}
+
+	deal, err := queryClient.GetDeal(dealID)
+	if err != nil {
+		log.Errorf("failed to get deal(%d): %s", dealID, err.Error())
+		http.Error(w, "failed to get deal", http.StatusBadRequest)
+		return
+	}
+
+	if err := validation.ValidateJSONSchemata(decryptedData, deal.DataSchema); err != nil {
+		log.Errorf("failed to validate data: %s", err.Error())
+		http.Error(w, "failed to validate data", http.StatusBadRequest)
+		return
+	}
+
+	// Re-encrypt data using a combined key
+	combinedKey := getCombinedKey(oraclePrivKey.Serialize(), dealID, dataHash)
+	reEncryptedData, err := crypto.Encrypt(combinedKey[:], nil, decryptedData)
+	if err != nil {
+		log.Errorf("failed to re-encrypt data with the combined key: %s", err.Error())
+		http.Error(w, "failed to re-encrypt data with the combined key", http.StatusInternalServerError)
+		return
+	}
+
+	// Put data into IPFS
+	cid, err := svc.IPFS().Add(reEncryptedData)
+	if err != nil {
+		log.Errorf("failed to store data to IPFS: %s", err.Error())
+		http.Error(w, "failed to store data to IPFS", http.StatusInternalServerError)
+		return
+	}
+
+	// Issue a certificate to the client
+	unsignedDataCert := &datadealtypes.UnsignedCertificate{
```

Comment:
I think we need to the `uniqueID` parameter.
However, it is not supported by the current panacea-core version.

You can add a unique ID later. :)

Reply:
Right. I'll add it after some PRs in panacea-core are merged